### DISCUSSION TITLE
Transaction management

### DISCUSSION THREAD

**Discussion by SFarhanizade** [link-intersystems/clean-architecture-example#3] [General]
Hi, I have a question about transaction management in clean architecture.
For example, in [this method](https://github.com/link-intersystems/clean-architecture-example/blob/c3c5dfa1c1159d69cf9ca6b07d8980525c8cbb9e/car-booking/use-cases/src/main/java/com/link_intersystems/carrental/booking/CarBookingInteractor.java#L27) you first ensured if the car is available, then you booked the car if everything is Okay, but there is a transactional problem in here. What if another request books the car at the same time? One of the requests may fail or the data may become corrupted. I want to know that how can you solve this problem with transaction management in clean architecture.
By the way, thank you for this great repository.

**Comment by Delis267** [comment]
He decided to wrap every use case with a transaction via [AOP](https://github.com/link-intersystems/clean-architecture-example/blob/main/app/app-components/car-rental-app-components-spi/src/main/java/com/link_intersystems/carrental/booking/CarBookingComponent.java#L23)

Both the [JDBC](https://github.com/link-intersystems/clean-architecture-example/blob/main/app/app-components/car-rental-app-components-jdbc/src/main/java/com/link_intersystems/carrental/components/jdbc/JdbcComponentsConfig.java#L39) and [JPA](https://github.com/link-intersystems/clean-architecture-example/blob/main/app/app-components/car-rental-app-components-jpa/src/main/java/com/link_intersystems/carrental/components/jpa/JpaComponentsConfig.java?utm_source=chatgpt.com) implementations provide their own transaction management setup. The [ComponentsConfigProvider](https://github.com/link-intersystems/clean-architecture-example/blob/main/app/app-components/car-rental-app-components-spi/src/main/java/com/link_intersystems/carrental/components/ComponentsConfigProvider.java?utm_source=chatgpt.com) loads the concrete implementation based on the provider available on the classpath.
